### PR TITLE
 [Feature] Support training on a subset of a dataset's classes

### PR DESCRIPTION
**Pull request by gboeer**
This PR introduces a new feature as discussed in #828 

**What**: Add `classes=` to train()/val()/predict() (Python API + CLI) so a dataset's declared classes can be filtered to a subset without editing annotation files.

**Why**: Users want to fine-tune on fewer classes than a dataset declares, without hand-editing labels or losing comparability with the original class ids.

- `classes=[...]` filters which original class ids receive supervision; kept ids are NOT compacted to a contiguous range, so predictions/checkpoints stay directly comparable to the full dataset (generalizes the existing `single_cls` remap-at-parse-time mechanism)
- `nc`/`names` are left untouched by `classes=`; only which boxes reach the loss changes
- gated to G0/G1 detection families (single_cls=True and classes=[...] both trigger the same gate)
- `model.val()` and the checkpoint resume path auto-inherit `classes` from the saved training config, same pattern as `single_cls`
- CLI: `libreyolo train ... classes=0,3,5` / `--classes 0,3,5`
- fixes: per-epoch training-time validation wasn't threading `classes=` into its own `ValidationConfig`, so periodic mAP was silently scored against the full, unfiltered dataset
- logs a clear warning when a run is training/validating on a class subset

## Code provenance
Original code written for this PR; bug fixes to LibreYOLO's own first-party code; no third-party code ported, adapted, or introduced; no GPL/AGPL/LGPL/non-commercial/unknown-license material involved.

<h2><a href="https://app.greptile.com/api/retrigger?id=64988894"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/RetriggerDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/Retrigger.svg?v=2">Retrigger</picture></a>Confidence Score: 1/5</h2>

The PR is not yet safe to merge because four previously reported behavioral gaps remain, including incomplete subset metrics and training paths that still ignore or cannot safely accept the option.

<h2><a href="https://app.greptile.com/api/ide/cursor?prompt=%23%23%23%20Issue%201%0Alibreyolo%2Fvalidation%2Fdetection_validator.py%3Aundefined-339%0AValidation%20filters%20the%20dataloader%20annotations%20by%20%60classes%60%2C%20but%20%60_init_metrics%60%20separately%20builds%20COCO%20ground%20truth%20from%20the%20full%20JSON%20or%20YOLO%20labels%2C%20and%20%60_update_metrics%60%20submits%20predictions%20without%20filtering%20their%20classes.%20As%20a%20result%2C%20standalone%20and%20per-epoch%20validation%20report%20mAP%20over%20the%20full%20dataset%20rather%20than%20the%20requested%20subset.%20With%20native%20COCO%20data%2C%20excluded%20labels%20can%20also%20fall%20through%20the%20incomplete%20label-to-category%20mapping.%20Apply%20the%20selected%20original%20class%20IDs%20consistently%20to%20evaluator%20ground%20truth%20and%20predictions.%0A%0A%23%23%23%20Issue%202%0Alibreyolo%2Ftraining%2Ftrainer.py%3A848-853%0AWhen%20a%20dataset%20uses%20an%20%60input_profile%60%2C%20this%20branch%20derives%20%60_class_remap%60%20and%20then%20immediately%20delegates%20to%20%60setup_histogram_data%60.%20That%20function%20creates%20%60YOLODataset%60%20without%20the%20remap.%20A%20supported%20G0%2FG1%20run%20therefore%20accepts%20and%20saves%20%60classes%60%2C%20but%20histogram%20training%20still%20supervises%20every%20annotation%20class.%0A%0A%23%23%23%20Issue%203%0Alibreyolo%2Fvalidation%2Fconfig.py%3A55-59%0A%60ValidationConfig%60%20adds%20%60classes%60%20without%20the%20normalization%20and%20validation%20already%20used%20by%20%60TrainConfig%60.%20Consequently%2C%20%60model.val%28classes%3D%5B%5D%29%60%20or%20a%20malformed%20checkpoint%20value%20can%20create%20an%20empty%20or%20negative%20remap%20and%20silently%20discard%20all%20dataloader%20labels%2C%20while%20the%20comma-separated%20form%20%60%220%2C3%2C5%22%60%20fails%20during%20integer%20conversion.%20Apply%20the%20same%20string%20normalization%20and%20non-empty%2C%20non-negative%2C%20duplicate%20checks%20at%20validation%20entry%20points.%0A%0A%23%23%23%20Issue%204%0Alibreyolo%2Fcli%2Fcommands%2Ftrain.py%3A323-328%0AOnly%20the%20train%20CLI%20exposes%20the%20new%20dataset-subset%20option.%20%60libreyolo%20val%20...%20classes%3D0%2C3%2C5%60%20is%20still%20rejected%20because%20%60val_cmd%60%20declares%20no%20%60classes%60%20option%20and%20does%20not%20pass%20one%20to%20%60loaded_model.val%60.%20Checkpoint%20inheritance%20does%20not%20address%20users%20who%20explicitly%20want%20to%20evaluate%20a%20different%20subset.%0A%0A%23%23%23%20Issue%205%0Atests%2Funit%2Ftest_classes_subset.py%3Aundefined-680%0AThis%20fast-unit%20test%20constructs%20a%20real%20LibreYOLO9%20model%2C%20runs%20a%20complete%20training%20epoch%2C%20and%20writes%20a%20checkpoint.%20That%20adds%20integration-level%20runtime%20and%20resource%20sensitivity%20to%20routine%20unit%20execution.%20Move%20this%20coverage%20to%20an%20integration%20or%20smoke%20tier%20and%20retain%20a%20lightweight%20mocked%20API%20test%20here.%0A%0ANote%3A%20If%20this%20suggestion%20doesn't%20match%20your%20team's%20coding%20style%2C%20reply%20to%20this%20and%20let%20me%20know.%20I'll%20remember%20it%20for%20next%20time!%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&pr=875&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=7"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=7">Fix All in Cursor</picture></a>Findings</h2>

1. P1&nbsp;**Subset Metrics Stay Unfiltered** <a href="https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701667">▶</a>
2. P1&nbsp;**Histogram Training Ignores Classes** <a href="https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701685">▶</a>
3. P1&nbsp;**Validation Classes Lack Checks** <a href="https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701700">▶</a>
4. P1&nbsp;**Validation CLI Lacks Classes** <a href="https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701717">▶</a>
5. P2&nbsp;**Unit Test Trains Model** <a href="https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701724">▶</a>

<details><summary>Summary</summary>

This PR adds class-subset filtering for G0/G1 detection training and validation while preserving original class IDs and the dataset’s full class metadata.
- Threads `classes` through configuration, dataset parsing, training, checkpoint resume, and periodic validation.
- Filters YOLO and native COCO ground truth for subset evaluation.
- Adds warnings and local test coverage for subset behavior.
</details>

<details open><summary>Diagram</summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
    A[classes configuration] --> B[Build original-ID remap]
    B --> C[Training dataset]
    B --> D[Validation dataset]
    C --> E[Drop excluded annotations]
    D --> F[Filter evaluator ground truth]
    G[Model predictions] --> H[COCO evaluator]
    F --> H
```
</details>

<sub>Reviews (2) · Last reviewed commit: ["fix:  Subset Metrics Stay Unfiltered"](https://github.com/libreyolo/libreyolo/commit/d68dd7246de1afee86b731173d423a6ead667f19)</sub>

### CODE REVIEW

**Comment by EHxuban11** [comment]
Hello @gboeer, thanks for the follow up to the issue creating this PR.

The overal approach is right but greptile flagged a few things which can be addressed in the next pass of code 🪖

 - Histogram training ignores classes (https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701685)
  - Validation needs checks for classes (https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701700)
  - The validation CLI is missing classes (https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701717)
  - Move the training test out of the unit suite (https://github.com/LibreYOLO/libreyolo/pull/875#discussion_r4026701724)

File: tests/unit/test_classes_subset.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+# Confirms the three places that used to independently resolve nc/names
+# (Trainer._resolve_num_classes_from_data_config, BaseTrainer._setup_data,
+# and YOLO9's own model.py::train()) agree: none of them need to know about
+# classes= at all anymore, since it never changes nc/names. That is what
+# makes this design change smaller and safer than the compacting-remap one
+# it replaces -- see build_class_remap's docstring.
+# ---------------------------------------------------------------------------
+
+
+def test_libreyolo9_train_end_to_end_keeps_original_nc_and_names(tmp_path):
```
[the remark this answers is not in the document]
Comment (gboeer) [reply tests/unit/test_classes_subset.py:769]: I thinkj @EHxuban11  should check if this is really a valid guideline. 

As far as i can see it, tests/unit/ already has an extensive, established convention of real (scratch-weights, CPU-only, tiny) training smoke tests living directly in the unit tier: test_yolo9_e2e_trainer_smoke.py, test_deim_trainer_smoke.py, test_dfine_sync_reduction.py.
Also, in docs/testing.md's own Scope line for the Unit tier: "CPU-safe behavior, config, parsing, errors, serialization, and CLI/API logic", nothing there excludes a tiny scratch-model training call.

File: libreyolo/training/trainer.py [hunk 2]
```diff
@@ -1515,8 +1603,9 @@ def _sync_wrapped_model_num_classes(self, num_classes: int) -> None:
                 f"{self.get_model_family()} wrapper rebuild did not sync the model "
                 f"head to num_classes={num_classes}; got {rebuilt_nc}."
             )
-        if self.config.single_cls:
-            wrapper.names = {0: "object"}
+        effective_names = self._effective_names_for_sync()
+        if effective_names is not None:
+            wrapper.names = effective_names
```
[the remark this answers is not in the document]
Comment (gboeer) [reply libreyolo/training/trainer.py:1616]: The two bugs share the exact same mechanism and the fix can't be meaningfully separated without leaving the reported symptom half-fixed, it would require overly complicated git magic for this case so I would like to ask for an exception to the one-problem-per-PR rule here.

[the remark this answers is not in the document]
Comment (gboeer) [reply libreyolo/training/trainer.py:1616]: With the latest commit, ive implemented the narrower gate, agreeing with the compromise:

Change to _effective_names_for_sync() (libreyolo/training/trainer.py): added a guard at the top, if not (self.config.single_cls or self.config.classes): return None, so name restoration only fires for the two cases this PR actually promises consistent metadata for. A plain full-dataset run that happens to trigger a head rebuild (e.g. resuming a checkpoint with a different nc) now goes back to the old behavior: _rebuild_for_new_classes()'s generic class_N placeholders stand, unfixed by this PR. Docstring updated to say so explicitly, including that fixing that case belongs in its own, separate change.
